pkg/southbound/gnmiClient.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2019-present Open Networking Foundation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package southbound
+
+import (
+	"context"
+	"github.com/openconfig/gnmi/client"
+	gclient "github.com/openconfig/gnmi/client/gnmi"
+	gpb "github.com/openconfig/gnmi/proto/gnmi"
+)
+
+// NewCaller :
+var NewCaller = func(ctx context.Context, d client.Destination) (GnmiClient, error) {
+	c, err := gclient.New(ctx, d)
+	return GnmiClientImpl{
+		c:            c,
+		capabilities: c.(*gclient.Client).Capabilities,
+		get:          c.(*gclient.Client).Get,
+		set:          c.(*gclient.Client).Set,
+		subscribe:    c.(*gclient.Client).Subscribe,
+	}, err
+}
+
+// GnmiClient :
+type GnmiClient interface {
+	Capabilities(ctx context.Context, r *gpb.CapabilityRequest) (*gpb.CapabilityResponse, error)
+	Get(ctx context.Context, r *gpb.GetRequest) (*gpb.GetResponse, error)
+	Set(ctx context.Context, r *gpb.SetRequest) (*gpb.SetResponse, error)
+	Subscribe(ctx context.Context, q client.Query) error
+	Recv() error
+	Close() error
+	Poll() error
+}
+
+// GnmiClientImpl :
+type GnmiClientImpl struct {
+	c            client.Impl
+	get          func(ctx context.Context, r *gpb.GetRequest) (*gpb.GetResponse, error)
```

Comment:
It's a nifty design alright. I see the advantage for substitution during testing. I'm not a fan of functions as variables in general (just an opinion). If there are other advantages to this (other than test) then it might be worthwhile going this direction.

Reply:
I think those function pointers might be completely redundant now - I'll play around a little with refactoring it.

Reply:
I'll push a change soon that removes these function pointers. Good catch! The only place that I absolutely need function pointers is in clientManager.go where it has a factory method to allocate the clients. Tests have to be able to override that from the outside. I'm going to refactor that code too, to make it clearer what is going on.

Reply:
I've also moved the factory function pointers out of clientManager into the individual client shim classes. This means less pollution for the manager implementation class.